clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,406 @@
+  private GravitinoClient client;
+  private String metalakeName;
+  private final ReentrantReadWriteLock rwLock = new ReentrantReadWriteLock();
+  private Cache<NameIdentifier, FilesetMeta> filesetCache;
+
+  @Override
+  public void initialize(URI name, Configuration configuration) throws IOException {
+    if (!name.toString().startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Unsupported file system scheme: %s for %s.",
+              name.getScheme(), GravitinoVirtualFileSystemConfiguration.GVFS_SCHEME));
+    }
+
+    int maxCapacity =
+        configuration.getInt(
+            GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_FILESET_CACHE_MAX_CAPACITY_KEY,
+            GravitinoVirtualFileSystemConfiguration
+                .FS_GRAVITINO_FILESET_CACHE_MAX_CAPACITY_DEFAULT);
+    Preconditions.checkArgument(maxCapacity > 0, "Cache max capacity should be greater than 0");
+
+    long evictionMillsAfterAccess =
+        configuration.getLong(
+            GravitinoVirtualFileSystemConfiguration
+                .FS_GRAVITINO_FILESET_CACHE_EVICTION_MILLS_AFTER_ACCESS_KEY,
+            GravitinoVirtualFileSystemConfiguration
+                .FS_GRAVITINO_FILESET_CACHE_EVICTION_MILLS_AFTER_ACCESS_DEFAULT);
+    Preconditions.checkArgument(
+        evictionMillsAfterAccess > 0, "Cache eviction mills after access should be greater than 0");
+
+    initializeCache(maxCapacity, evictionMillsAfterAccess);
+
+    // initialize the Gravitino client
+    String serverUri =
+        configuration.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_SERVER_URI_KEY);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(serverUri), "Gravitino server uri is not set in the configuration");
+
+    this.metalakeName =
+        configuration.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_CLIENT_METALAKE_KEY);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(metalakeName), "Gravitino metalake is not set in the configuration");
+
+    // TODO Need support more authentication types, now we only support simple auth
+    this.client =
+        GravitinoClient.builder(serverUri).withMetalake(metalakeName).withSimpleAuth().build();
+
+    // Close the gvfs cache to achieve tenant isolation based on different user tokens in the
+    // configuration.
+    configuration.set(
+        String.format(
+            "fs.%s.impl.disable.cache", GravitinoVirtualFileSystemConfiguration.GVFS_SCHEME),
+        "true");
+    setConf(configuration);
+
+    NameIdentifier filesetIdentifier = extractIdentifier(name);
+    getCachedFileset(filesetIdentifier);
+    this.workingDirectory = new Path(name);
+    this.uri = URI.create(name.getScheme() + "://" + name.getAuthority());
+    super.initialize(uri, getConf());
+  }
+
+  private void initializeCache(int maxCapacity, long expireAfterAccess) {
+    this.filesetCache =
```

Comment:
The cache here is effective for reusing a FileSystem. For example, the following usage:
```
Configuration conf = new Configuration();
conf.set("fs.gvfs.impl", "com.datastrato.gravitino.filesystem.hadoop.GravitinoVirtualFileSystem");
conf.set("fs.AbstractFileSystem.gvfs.impl", "com.datastrato.gravitino.filesystem.hadoop.Gvfs");
conf.set("fs.gravitino.server.uri", "http://localhost:8090");
conf.set("fs.gravitino.client.metalake", "gravitino_metalake");
conf.set("fs.gravitino.client.token", "xxxxxxx");

// User has the path1's access permission
Path path1 = new Path("gvfs://fileset/fileset_catalog/test_schema/fileset_1");
FileSystem fs = path1.getFileSystem();
FileStatus[] fileStatuses1 = fs.listStatus(path1);

// User can reuse the fs to access the path2 if they have the access permission
Path path2 = new Path("gvfs://fileset/fileset_catalog/test_schema/fileset_2");
FileStatus[] fileStatuses2 = fs.listStatus(path2);
```
In this case, if this user has the fileset permissions corresponding to `path1` and `path2`, he can use the same fs to access them, which will be saved in the cache here.
In addition, the cache here should be kept independent because what is stored here are fileset instances obtained through the server using the current user's token to authenticate. This cannot be shared between FileSystems.

Reply:
Thanks for your explain. Let me think about it twice.